crates/router/src/core/payments.rs

```diff
@@ -1177,20 +1184,24 @@ pub async fn list_payments(
     merchant: domain::MerchantAccount,
     constraints: api::PaymentListConstraints,
 ) -> RouterResponse<api::PaymentListResponse> {
-    use futures::stream::StreamExt;
-
     use crate::types::transformers::ForeignFrom;
 
     helpers::validate_payment_list_request(&constraints)?;
     let merchant_id = &merchant.merchant_id;
-    let payment_intents =
+    let mut payment_intents =
         helpers::filter_by_constraints(db, &constraints, merchant_id, merchant.storage_scheme)
             .await
             .to_not_found_response(errors::ApiErrorResponse::PaymentNotFound)?;
-
-    let pi = futures::stream::iter(payment_intents)
-        .filter_map(|pi| async {
-            let pa = db
+    payment_intents.sort();
+    payment_intents.reverse();
+
+    let payment_intent_attempt_map: Arc<RwLock<HashMap<&str, Vec<storage::PaymentAttempt>>>> =
+        Arc::new(RwLock::new(HashMap::new()));
```

Comment:
I don't think interior mutability is required here...

you could do something like
```rust
// psuedo code
let pi_pa_mapping = payment_intents.iter().map(|pi| async {
     let payment_attempt = get_attempt_from_intent(pi);
     (pi.payment_id, payment_attempt)
}).collect::<HashMap<String, PaymentAttempt>>;
let sorted_payment_attempts = payment_intents.iter().map(|pi| pi_pa_mapping.pop(pi.payment_id).collect::<Vec<PaymentAttempt>>;
```

While this isn't a performance sensitive part, locks generally have a considerable overhead,
& I think doing it without locks might lead to prettier code as well

Reply:
Ah yes! That looks much better. I did not know HashMaps could be collected (TIL). Ill look into sorting payment intents in diesel queries itself.

Reply:
A single payment intent can have many payment attempts right? If I were to collect a hashmap of <&str, PaymentAttempt> Id be overwriting and only storing last payment attempt.

Reply:
since you'll be fetching via `active_attempt_id` which is like a primary key on attempt id, you'll be fetching only 1 attempt per payment intent (which is the current active one)